python/cdp_montana_legislature_backend/scraper.py

```diff
@@ -36,5 +40,77 @@ def get_events(
     from GitHub Actions UI.
     """
 
-    # Your implementation here
+    # TODO: Read key bills from config or Capitol Tracker JSON
+    key_bill_names = ["HB 1"]
+    
+    # Start at the big table of all bills from the 2021 session.
+    bills_url_2021 = "http://laws.leg.mt.gov/legprd/LAW0217W$BAIV.return_all_bills?P_SESS=20211"
+    bills_html = requests.get(bills_url_2021).text
+    parsed_bills_html = BeautifulSoup(bills_html, 'html.parser')
+    bills_table = parsed_bills_html.find_all('table')[1]
+    bills_table_rows = bills_table.find_all('tr')
+    print(f"Found table with {len(bills_table_rows) - 1} rows.")
+
+    # Of all the bills, narrow down to only the key bills and store off the LAWS bill URL for the next step.
+    key_bills_data = []
+    for bill_row in bills_table_rows[1:]:
+        bill_link = bill_row.find_all('a')[0]
+        bill_type_number = bill_link.text
+        bill_data = {}
+        if bill_type_number in key_bill_names:
+            bill_data['bill_type_number'] = bill_type_number
+            bill_data['description'] = bill_row.find_all('td')[-1].text
+            bill_data['laws_bill_url'] = "http://laws.leg.mt.gov/legprd/" + bill_link['href']
+            key_bills_data.append(bill_data)
+
+    print(f"Found key bills: {key_bills_data}")
+
+    event_data = []
+    # Go to each LAWS bill URL and find bill actions that have associated recordings.
+    for bill_data in key_bills_data:
+        laws_bill_html = requests.get(bill_data['laws_bill_url']).text
+        bill_rows_with_recordings = re.findall('.*sliq.*', laws_bill_html)
+        for bill_row in bill_rows_with_recordings:
+            # TODO: Use from_dt and to_dt to filter bill actions within date range
+            parsed_bill_row = BeautifulSoup(bill_row, 'html.parser')
+            bill_cells = parsed_bill_row.find_all('td')
+            all_links = bill_cells[-1].find_all('a')
+            sliq_links = [ link for link in all_links if "sliq" in link['href'] ]
```

Comment:
BeautifulSoup also supports a parameter to find that you can pass an regex ... so instead of list comprehension here to filter the anchor tags, this becomes...

```
bill_cells[-1].find_all('a', href=re.compile('sliq'))
```

Reply:
oo la la 👍